bpo-43931: Export Python version as API data (#25577)
python/cpython: Doc/c-api/apiabiversion.rst

   Thus ``3.4.1a2`` is hexversion ``0x030401a2`` and ``3.10.0`` is
   hexversion ``0x030a00f0``.

   This version is also available via the symbol :data:`Py_Version`.

.. c:var:: const unsigned long Py_Version

>    The Python runtime version number encoded in a single constant integer, with
>    the same format as the c:macro:`PY_VERSION_HEX` macro.

Reviewer: I'd still like to make it more explicit that the note from the top of the file doesn't apply to `Py_Version`.

```suggestion
   The Python runtime version number encoded in a single constant integer, with
   the same format as the c:macro:`PY_VERSION_HEX` macro.
   This contains the Python version used at run time.
```

Author: I agree and that's why I've specified "_runtime_ version number" on line 65. Perhaps adding some _emphasis_ to it would be enough? Otherwise, I'll add the suggested line 67.

>    This version is also available via the symbol :data:`Py_Version`.
> 
> .. c:var:: const unsigned long Py_Version
> 
>    The Python version number encoded as the single constant integer, same value 
>    as the c:macro:`PY_VERSION_HEX` macro.
> 
>    .. versionadded:: 3.11

Reviewer: But they don't always have the same value – one is retrieved at runtime, the other is compiled into an extension.
```suggestion

.. c:var:: const unsigned long Py_Version

   The Python version number encoded as a single constant integer
   in the same format as the c:macro:`PY_VERSION_HEX` macro.
   This contains the Python version used at run time.

   .. versionadded:: 3.11
```

Reviewer 2: "same format as the PY_VERSION_HEX macro" sounds like a good description. So we only have to document the format once, in PY_VERSION_HEX.

Author: Hmm, I'm not sure I understand this. The `PY_VERSION_HEX` is a C macro, which is then not available at runtime,  and `Py_Version` is binary constant data, that is assigned the value of `PY_VERSION_HEX` at compile time. Loosely speaking, one can use `sys.hexversion` to access `PY_VERSION_HEX` from Python, and I can see why `sys.hexversion` might not be the same as `Py_Version`. But the documentation here is stating that `PY_VERSION_HEX` and `Py_Version` have the same value at the C API level, which should always be true?

Reviewer 2: I guess that Petr is talking about a C extension built with Python 3.8 and run with Python 3.11.

Py_Version and sys.hexversion is the version of the running Python process. PY_VERSION_HEX can be different in a C extension built with an older Python version.

I understood that the documentation can be misleading in such corner case.

Reviewer: Yes.
This is what the note at the top of [the document](https://docs.python.org/3.11/c-api/apiabiversion.html) is about:

> CPython exposes its version number in the following macros. Note that these correspond to the version code is **built with**, not necessarily the version used at **run time**.

Feel free to clarify it if you find better wording :)

Author: Ah, I see now, thanks for clarifying!

   Thus ``3.4.1a2`` is hexversion ``0x030401a2`` and ``3.10.0`` is
   hexversion ``0x030a00f0``.

   This version is also available via the symbol :data:`Py_Version`.

.. c:var:: const unsigned long Py_Version

   Exported symbol that provides the same version information as
   ``PY_VERSION_HEX``. Useful for determining the version of Python without

>    having to run it, or when it is embedded into other binaries.

Reviewer 2: Hum, I suggest:

```suggestion
   The Python version number encoded in a single integer, same value than
   the :macro:`PY_VERSION_HEX` macro but as a constant variable.
```

Once we agree on a description, we can reuse in in the What's New and NEWS entries.

Author: Done. I'm a bit reluctant to use `constant variable` though, as it sounds oxymoronic to me (something is either constant or variable). I hope that what I wrote is good too :slightly_smiling_face: .

   Thus ``3.4.1a2`` is hexversion ``0x030401a2`` and ``3.10.0`` is
   hexversion ``0x030a00f0``.

>    Starting with Python 3.11, this information is also available via the
>    exported symbol `Py_Version`.

Reviewer 2: There is no need to repeat the version change here.

```suggestion
   This version is also available via the symbol :data:`Py_Version`.
```

>    exported symbol `Py_Version`.

Reviewer 2: Can you please add a similar note in Py_GetVersion() documentation? See Doc/c-api/init.rst.

> .. c:var:: Py_Version

Reviewer 2: Doc build fails, you must specify the type:

```suggestion
.. c:var:: const long Py_Version
```

> .. c:var:: const long Py_Version

Reviewer 2: Is it useful to use a signed number? Why not using unsigned long?

   Thus ``3.4.1a2`` is hexversion ``0x030401a2`` and ``3.10.0`` is
   hexversion ``0x030a00f0``.

>    Starting with Python 3.11, this information is also available via the

Reviewer 2: Please document also the variable using the `.. c:var:: PY_VERSION_HEX` markup and add a `.. versionadded:: 3.11` markup.
